components/web_discovery/browser/server_config_loader.cc

```diff
@@ -0,0 +1,446 @@
+constexpr base::TimeDelta kPatternsRequestLatestDelay = base::Minutes(3);
+constexpr base::TimeDelta kPatternsRequestInitDelay = base::Seconds(15);
+
+constexpr size_t kPatternsMaxFileSize = 128000;
+
+constexpr net::NetworkTrafficAnnotationTag kNetworkTrafficAnnotation =
+    net::DefineNetworkTrafficAnnotation("wdp_config", R"(
+    semantics {
+      sender: "Brave Web Discovery Server Configuration Fetch"
+      description:
+        "Requests server configuration needed to send Web Discovery "
+        "measurements to Brave servers."
+      trigger:
+        "Requests are automatically sent at intervals "
+        "while Brave is running."
+      data: "Configuration attributes"
+      destination: WEBSITE
+    }
+    policy {
+      cookies_allowed: NO
+      setting:
+        "Users can opt-in or out via brave://settings/search"
+    })");
+
+constexpr char kGroupPubKeysFieldName[] = "groupPubKeys";
+constexpr char kPubKeysFieldName[] = "pubKeys";
+constexpr char kMinVersionFieldName[] = "minVersion";
+constexpr char kKeysFieldName[] = "keys";
+constexpr char kLimitFieldName[] = "limit";
+constexpr char kPeriodFieldName[] = "period";
+constexpr char kSourceMapFieldName[] = "sourceMap";
+constexpr char kSourceMapActionsFieldName[] = "actions";
+constexpr char kLocationFieldName[] = "location";
+
+constexpr char kCollectorConfigPathWithFields[] =
+    "/config?fields=minVersion,groupPubKeys,pubKeys,sourceMap";
+constexpr char kQuorumConfigPath[] = "/config";
+constexpr char kPatternsFilename[] = "wdp_patterns.json";
+
+constexpr char kOmittedLocationValue[] = "--";
+constexpr auto kAllowedReportLocations =
+    base::MakeFixedFlatSet<std::string_view>(
+        {"ar", "at", "au", "be", "br", "ca", "ch", "cn", "cz", "de",
+         "dk", "es", "fi", "fr", "gb", "gr", "hu", "in", "it", "jp",
+         "mx", "nl", "no", "pl", "ro", "ru", "se", "ua", "us"});
+
+KeyMap ParseKeys(const base::Value::Dict& encoded_keys) {
+  KeyMap map;
+  for (const auto [date, key_b64] : encoded_keys) {
+    std::vector<uint8_t> decoded_data;
+    // Decode to check for valid base64
+    if (!base::Base64Decode(key_b64.GetString())) {
+      continue;
+    }
+    map[date] = key_b64.GetString();
+  }
+  return map;
+}
+
+base::flat_map<std::string, std::unique_ptr<SourceMapActionConfig>>
+ParseSourceMapActionConfigs(const base::Value::Dict& configs_dict) {
+  base::flat_map<std::string, std::unique_ptr<SourceMapActionConfig>> map;
+  for (const auto [action, config_dict_val] : configs_dict) {
+    auto* config_dict = config_dict_val.GetIfDict();
+    if (!config_dict) {
+      continue;
+    }
+    auto& action_config = map[action];
+    if (!action_config) {
+      action_config = std::make_unique<SourceMapActionConfig>();
+    }
+    auto* keys_list = config_dict->FindList(kKeysFieldName);
+    if (keys_list) {
+      for (const auto& key_val : *keys_list) {
+        if (key_val.is_string()) {
+          action_config->keys.push_back(key_val.GetString());
```

Comment:
again, use non-const references and move everything you can.

Reply:
same question: https://github.com/brave/brave-core/pull/24969#discussion_r1708086771